src/containers/ListingPage/ListingPage.duck.js

```diff
@@ -185,13 +185,56 @@ export const fetchReviews = listingId => (dispatch, getState, sdk) => {
     });
 };
 
-export const fetchTimeSlots = params => (dispatch, getState, sdk) => {
+const timeSlotsRequest = params => (dispatch, getState, sdk) => {
+  return sdk.timeslots.query(params).then(response => {
+    return denormalisedResponseEntities(response);
+  });
+};
+
+export const fetchTimeSlots = listingId => (dispatch, getState, sdk) => {
   dispatch(fetchTimeSlotsRequest);
-  return sdk.timeslots
-    .query(params)
-    .then(response => {
-      const timeSlots = denormalisedResponseEntities(response);
-      dispatch(fetchTimeSlotsSuccess(timeSlots));
+
+  // Time slots can be fetched for 90 days at a time,
+  // for at most 180 days from now. If max number of bookable
+  // day exceeds 90, a second request is made.
+
+  const maxTimeSlots = 90;
+  // booking range: today + bookable days -1
+  const bookingRange = config.dayCountAvailableForBooking - 1;
+  const timeSlotsRange = Math.min(bookingRange, maxTimeSlots);
+
+  const start = moment
+    .utc()
+    .startOf('day')
+    .toDate();
+  const end = moment()
+    .utc()
+    .startOf('day')
+    .add(timeSlotsRange, 'days')
+    .toDate();
+  const params = { listingId, start, end };
+
+  return dispatch(timeSlotsRequest(params))
+    .then(timeSlots => {
+      const secondRequest = bookingRange > maxTimeSlots;
+
+      if (secondRequest) {
+        const secondRange = Math.min(maxTimeSlots, bookingRange - maxTimeSlots);
+        const secondParams = {
+          listingId,
+          start: end,
+          end: moment(end)
+            .add(secondRange, 'days')
+            .toDate(),
+        };
+
+        dispatch(timeSlotsRequest(secondParams)).then(secondBatch => {
+          const combined = timeSlots.concat(secondBatch);
+          dispatch(fetchTimeSlotsSuccess(combined));
+        });
```

Comment:
Should there be a `return` before the `dispatch` or a `.catch` on the promise to handle error in the request?

Reply:
The `.catch` is left out so that in case of an error the `.catch` of the outer promise handles it. This way there error action is fired in case either of the time slot requests fail. A succeed action is dispatched when both requests succeed.

Reply:
How would a `return` help with handling a failing request here?

Reply:
The failing promise never goes to the outer `.catch` without the `return`.

Reply:
Well so it seems. I was sure I had this tested. Well spotted 👍